packages/smooth_app/lib/pages/folksonomy/folksonomy_manager.dart

```diff
@@ -0,0 +1,260 @@
+import 'package:smooth_app/database/dao_folksonomy.dart';
+import 'package:smooth_app/database/dao_transient_folksonomy.dart';
+import 'package:smooth_app/database/local_database.dart';
+import 'package:smooth_app/pages/folksonomy/folksonomy_provider.dart';
+import 'package:smooth_app/query/product_query.dart';
+
+/// Writer class around Folksonomy, temporary until we implement BackgroundTasks
+class FolksonomyManager {
+  FolksonomyManager(this._localDatabase);
+
+  final LocalDatabase _localDatabase;
+  DaoFolksonomy get _daoFolksonomy => DaoFolksonomy(_localDatabase);
+  DaoTransientFolksonomy get _daoTransientFolksonomy =>
+      DaoTransientFolksonomy(_localDatabase);
+
+  String? _bearerToken;
+
+  Future<void> addTag(String barcode, String key, String value) async {
+    final List<ProductTag> currentTags =
+        await _daoFolksonomy.get(barcode) ?? <ProductTag>[];
+    final bool exists = currentTags.any((ProductTag tag) => tag.key == key);
+    if (exists) {
+      throw Exception('This tag already exists!');
+    }
+
+    final ProductTag newProductTag = ProductTag(
+      barcode: barcode,
+      key: key,
+      value: value,
+      owner: '',
+      version: 1,
+      editor: '',
+      lastEdit: DateTime.now(),
+      comment: '',
+    );
+
+    await _daoTransientFolksonomy.put(barcode, <FolksonomyOperation>[
+      ...getPendingOperationsByBarcode(barcode),
+      FolksonomyOperation(type: FolksonomyAction.add, tag: newProductTag),
+    ]);
+    unawaited(syncProductTags(barcode));
+  }
+
+  Future<void> editTag(
+    String barcode,
+    String key,
+    String newValue,
+    int newVersion,
+  ) async {
+    final ProductTag editedProductTag = ProductTag(
+      barcode: barcode,
+      key: key,
+      value: newValue,
+      owner: '',
+      version: newVersion,
+      editor: '',
+      lastEdit: DateTime.now(),
+      comment: '',
+    );
+
+    await _daoTransientFolksonomy.put(barcode, <FolksonomyOperation>[
+      ...getPendingOperationsByBarcode(barcode),
+      FolksonomyOperation(type: FolksonomyAction.edit, tag: editedProductTag),
+    ]);
+    unawaited(syncProductTags(barcode));
+  }
+
+  Future<void> deleteTag(String barcode, String key, int version) async {
+    final ProductTag tagToDelete = ProductTag(
+      barcode: barcode,
+      key: key,
+      value: '',
+      owner: '',
+      version: version,
+      editor: '',
+      lastEdit: DateTime.now(),
+      comment: '',
+    );
+
+    await _daoTransientFolksonomy.put(barcode, <FolksonomyOperation>[
+      ...getPendingOperationsByBarcode(barcode),
+      FolksonomyOperation(type: FolksonomyAction.remove, tag: tagToDelete),
+    ]);
+    unawaited(syncProductTags(barcode));
+  }
+
+  Future<void> syncAllProductTags() async {
+    final List<String> barcodes = _daoTransientFolksonomy.getAllBarcodes();
+    for (final String barcode in barcodes) {
+      await syncProductTags(barcode);
+    }
+  }
+
+  Future<void> syncProductTags(String barcode) async {
```

Comment:
```suggestion
  Future<void> serverPerformActions(String barcode) async {
```

The choice of `sync` is confusing.